src/program/latex/mod.rs

```diff
@@ -0,0 +1,1530 @@
+    gates: HashMap<u32, String>,
+    /// control at column with distance from targ wire
+    ctrl: HashMap<u32, i64>,
+    /// at this column is the wire a target?
+    targ: HashMap<u32, bool>,
+    /// any parameters required at column on the wire for gates
+    parameters: HashMap<u32, Vec<Parameter>>,
+    /// any modifiers added to the wire at column
+    modifiers: HashMap<u32, Vec<String>>,
+    /// empty column
+    empty: HashMap<u32, Command>,
+}
+
+impl Default for Wire {
+    fn default() -> Self {
+        Self {
+            name: 0,
+            gates: HashMap::new(),
+            ctrl: HashMap::new(),
+            targ: HashMap::new(),
+            parameters: HashMap::new(),
+            modifiers: HashMap::new(),
+            empty: HashMap::new(),
+        }
+    }
+}
+
+impl Wire {
+    /// Retrieves a gate's parameters from Expression and matches them with its
+    /// symbolic definition which is then stored into wire at the specific
+    /// column.
+    ///
+    /// # Arguments
+    /// `&mut self` - exposes the Wire's parameters at this column
+    /// `expression` - expression from Program to get name of parameter
+    /// `column` - the column taking the parameters
+    /// `texify` - is texify_numerical_constants setting on?
+    fn set_param(&mut self, expression: &Expression, column: u32, texify: bool) {
+        let text: String;
+
+        // get the name of the supported expression
+        match expression {
+            Expression::Address(mr) => {
+                text = mr.name.to_string();
+            }
+            Expression::Number(c) => {
+                text = c.re.to_string();
+            }
+            expression => text = expression.to_string(),
+        }
+
+        let param;
+        // if texify_numerical_constants
+        if texify {
+            // get the matching symbol from text
+            param = vec![Parameter::Symbol(Symbol::match_symbol(text))];
+        } else {
+            // set the symbol as text
+            param = vec![Parameter::Symbol(Symbol::Text(text))];
+        }
+        self.parameters.insert(column, param);
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum LatexGenError {
+    #[error("Cannot parse LaTeX for unsupported program: {program}")]
+    UnsupportedProgram { program: String },
+    #[error("Cannot parse LaTeX for unsupported gate: {gate}")]
+    UnsupportedGate { gate: String },
+    #[error("Tried to parse CNOT and found a control qubit without a target.")]
+    FoundCNOTWithNoTarget,
+}
+
+/// Supported types
+enum Supported {
+    Gate(SupportedGate),
+    New,  // New initialization of a non-None variant
+    None, // ()
+}
+
+/// Set of all Gates that can be parsed to LaTeX
+#[derive(PartialEq)]
+enum SupportedGate {
+    Pauli(String),
+    Hadamard(String),
+    Phase(String),
+    ControlledPhase(String),
+    ControlledX(String),
+    DefGate(String),
+    Modifiers(String),
+    Unsupported(String), // for error handling
```

Comment:
If this has an `Unsupported` variant, the `enum` should not be called `SupportedGate`